File: audience/src/main/java/br/com/concretesolutions/audience/core/concurrent/MessageDispatcher.java
package br.com.concretesolutions.audience.core.concurrent;

import android.os.Message;

import java.util.Random;

import br.com.concretesolutions.audience.core.Director;
import br.com.concretesolutions.audience.core.MessageEvent;

public final class MessageDispatcher {

    private static final int N_THREADS = 4;

    // Handlers --------------------
    private final BaseHandler[] schedulers = new BaseHandler[N_THREADS];
    private final UIInboxHandler uiScheduler = new UIInboxHandler();

    // Queue selection -------------
    private final Random mQueueSelectionRandom = new Random();

    public MessageDispatcher() {
        // Initialize system
        for (int i = 0; i < N_THREADS; i++) {
            final ActorThread actorThread = new ActorThread();
            actorThread.start();
            schedulers[i] = actorThread.getMessageHandler();
        }
    }

    public void shutdown() {
        for (BaseHandler handler : schedulers)
            handler.getLooper().quit();
    }

    public <T> void send(MessageEvent<T> ref) {
        final BaseHandler scheduler = ref.isOnStage ? uiScheduler : getNextQueue();
        final Message message = scheduler.obtainMessage(0, ref);
        scheduler.sendMessage(message);
    }

    private BaseHandler getNextQueue() {
        return schedulers[mQueueSelectionRandom.nextInt(N_THREADS)];
    }
}

File: audience-core/src/main/kotlin/io/github/victorolinasc/audience/core/concurrent/MessageDispatcher.kt
package io.github.victorolinasc.audience.core.concurrent

import io.github.victorolinasc.audience.core.MessageEvent
import java.util.*

class MessageDispatcher {

    // Handlers --------------------
    private val schedulers = arrayOfNulls<BaseHandler>(N_THREADS)
    private val uiScheduler = UIInboxHandler()

    // Queue selection -------------
    private val mQueueSelectionRandom = Random()

    private val nextQueue: BaseHandler
        get() = schedulers[mQueueSelectionRandom.nextInt(N_THREADS)]!!

    init {
        // Initialize system
        for (i in 0 until N_THREADS) {
            val actorThread = ActorThread()
            actorThread.start()
            schedulers[i] = actorThread.messageHandler
        }
    }

    fun shutdown() {
        for (handler in schedulers)
            handler!!.looper.quit()
    }

    fun <T> send(ref: MessageEvent<T>) {
        val scheduler = if (ref.isOnStage) uiScheduler else nextQueue
        val message = scheduler.obtainMessage(0, ref)
        scheduler.sendMessage(message)
    }

    companion object {
        private const val N_THREADS = 4
    }
}
